Log warning when using the default json settings

timbussmann opened the pull request:
Small change that brings more attention to users about the [default `TypeNameHandling` settings](https://docs.particular.net/nservicebus/serialization/newtonsoft#typenamehandling) via logging a warning when running with the default settings (using `TypeNameHandling.Auto`). This log warning disappears when the user provides a custom setting that either explicitly enables `TypeNameHandling.Auto` or uses the default `TypeNameHandling` setting which is `None`. Since this doesn't change the default behavior itself, this is not a breaking change unless the user makes the explicit decision to change the settings.

I'd prefer to not call logging in the ctor of the settings but I think it's acceptable in this case given all the default handling is happening in the `JsonMessageSerializer` ctor.

danielmarbach commented on src/NServiceBus.Newtonsoft.Json/JsonMessageSerializer.cs:
```diff
@@ -26,10 +27,11 @@ class JsonMessageSerializer : IMessageSerializer
         {
             this.messageMapper = messageMapper;
 
-            settings ??= new JsonSerializerSettings
```
Would it be better to still use the default ctor and then check `settings.TypeNameHandling == TypeNameHandling.Auto`? The benefit of this would be that if at any time they change the default in newtonsoft we would still be using the default setting (whatever that is at the point in time).

I'm mentioning it because at this stage our implemention is:

When no setting provided use the default settings of the version we bound

your PR would change that though to

Whe no setting provided use the default setting exception set the `TypeNameHandling` to always be `Auto`.

timbussmann replied:
> The benefit of this would be that if at any time they change the default in newtonsoft we would still be using the default setting (whatever that is at the point in time).

I'm pretty sure they will never change that default in Json.NET and also we don't want to adopt changing defaults automatically either as that would be compatiblity-breaking change.

> When no setting provided use the default settings of the version we bound

not sure what you mean by that? The current implementation is "use the default settings but set TypeNameHandling to Auto".

danielmarbach replied:
My point was that the current design of the newtonsoft implementation is that it takes the package version default behavior of the default constructor of the JsonSettings. The change here does multiple things:

- It changes that design
- It introduces backward compatibility by hardcoding the auto-handling
- It logs a warning

That's why I brought this up to clarify if we really should couple those things together in one go. As long as we have thought about the implications, I'm fine. I was just thinking that if we want to not change the previous design assumptions, staying with newing up the default settings and logging when type name handling is set to auto would achieve the goal of making the users aware.

danielmarbach replied:
OH duh, I completely missed that we set it before https://github.com/Particular/NServiceBus.Newtonsoft.Json/pull/215/files#diff-f83cd6aed4a4f832e85cf6e172f0586133dd3318b269fd9d226a9c2bbaac0f72L31 in the diff. Never mind my comment. Sorry for holding things up. I should have looked closer

danielmarbach commented on src/NServiceBus.Newtonsoft.Json/JsonMessageSerializer.cs:
```diff
@@ -26,10 +27,11 @@ class JsonMessageSerializer : IMessageSerializer
-                TypeNameHandling = TypeNameHandling.Auto
-            };
+                settings = new JsonSerializerSettings { TypeNameHandling = TypeNameHandling.Auto };
+                log.Warn($"The default {nameof(JsonSerializerSettings)} for NServiceBus.Newtonsoft.Json use TypeNameHandling.Auto for backwards compatibility. This is a potential security vulnerability and it is recommended to use TypeNameHandling.None if possible. To disable this warning, provide a custom {nameof(JsonSerializerSettings)} instance to 'endpointConfiguration.UseSerialization<Newton ...
```
Ah this is super cool

https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca2328
https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca2327

Should we mention those links in the warning as a reference? Or refer to a docs site we own that mentiones those analyzer rules as a possible mitigation strategy?

On another note, the above links also mention the serializer binder being an issue. Have we thought about checking that too?

timbussmann replied:
those rules don't really show up when using our newtonsoft json package, so I'm not sure why we would link to them other than as a justification why this is a bad default?

We do not check the binders as we use the default settings in that regards and afaik the default settings are safe.

danielmarbach replied:
I was thinking about an angle to not just say "it is bad" but actually link to information why it is bad.

timbussmann replied:
I wanted to keep the log message a bit on the compact side, and I think it's probably not the best idea to link to external resources from within our log messages. But we could link to our own docs and add more documentation there.

danielmarbach replied:
> Or refer to a docs site we own that mentiones those analyzer rules as a possible mitigation strategy?

Yep I also suggested that. I agree external links are less preferrable

ramonsmits commented on src/NServiceBus.Newtonsoft.Json/JsonMessageSerializer.cs:
```diff
@@ -26,10 +27,11 @@ class JsonMessageSerializer : IMessageSerializer
-                TypeNameHandling = TypeNameHandling.Auto
-            };
+                settings = new JsonSerializerSettings { TypeNameHandling = TypeNameHandling.Auto };
+                log.Warn($"The default {nameof(JsonSerializerSettings)} for NServiceBus.Newtonsoft.Json use TypeNameHandling.Auto for backwards compatibility. This is a potential security vulnerability and it is recommended to use TypeNameHandling.None if possible. To disable this warning, provide a custom {nameof(JsonSerializerSettings)} instance to 'endpointConfiguration.UseSerialization<Newton ...
```
Should be link to docs search?
```suggestion
                log.Warn($"The default {nameof(JsonSerializerSettings)} for NServiceBus.Newtonsoft.Json uses TypeNameHandling.Auto for backwards compatibility. This is a potential security vulnerability and it is recommended to use TypeNameHandling.None if possible. To disable this warning, provide a custom {nameof(JsonSerializerSettings)} instance to 'endpointConfiguration.UseSerialization<NewtonsoftSerializer>().Settings'. See https://docs.particular.net/search?q=newtonsoftjson+typenamehandling");
```

timbussmann replied:
I think we usually don't hardcode links in exception messages as this might be a problem in the long run.

ramonsmits replied:
@timbussmann That is why it links to search and not a specific page. It occasionally happens, see https://github.com/search?l=C%23&q=org%3AParticular+https%3A%2F%2Fdocs.particular.net%2Fsearch&type=Code

timbussmann commented:
@danielmarbach @ramonsmits I added another sentence to the log message that refers to the documentation more explicitly. I've added a link to the main docs URL as I think this URL should remain stable with minimal risks. I'm proposing some more details on TypeNameHandling to the docs via this PR: https://github.com/Particular/docs.particular.net/pull/5603
